File: trunk/csharp/Nase/Parser.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Nase.Syntax;

namespace Nase
{
    class Parser
    {
        static readonly Logger Logger = LogManager.CreateLogger();

        FileManager _fileManager;
        Scanner _scanner;

        public Parser(FileManager fileManager, Scanner scanner)
        {
            this._fileManager = fileManager;
            this._scanner = scanner;
        }

        public bool ParseProgramm(out SyntaxTree syntaxTree)
        {
            syntaxTree = null;

            SyntaxTreeNode rootNode = IsProgramm();
            syntaxTree = new SyntaxTree(rootNode);

            Logger.Debug(this._scanner.SymbolTable.DumpSymbolTable());
            Logger.Debug(syntaxTree.DumpTreeTable());

            return rootNode != null;
        }

        FilePosition GetInputFilePosition()
        {
            return this._fileManager.Input.Position;
        }

        /*
         * program = statementSequence EOF_SYMBOL.
         */
        SyntaxTreeNode IsProgramm()
        {
            SyntaxTreeNode statementSequenceNode;

            FilePosition position = GetInputFilePosition();
            if (null != (statementSequenceNode = IsStatementSequence()))
            {
                if (IsEofSymbol())
                {
                    return new SyntaxTreeProgramNode(position, statementSequenceNode);
                }
            }
            return null;
        }

        /*
         * statementSequence = statement { statement }.
         */
        SyntaxTreeNode IsStatementSequence()
        {
            SyntaxTreeNode statementSequenceNode = null;
            SyntaxTreeNode statementNode;

            FilePosition position = GetInputFilePosition();
            if (null != (statementNode = IsStatement()))
            {
                statementSequenceNode = new SyntaxTreeSequenceNode(position, statementNode, statementSequenceNode);
            }

            SyntaxTreeNode firstStatementNode = statementSequenceNode;

            while (null != (statementNode = IsStatement()))
            {
                position = GetInputFilePosition();
                statementSequenceNode = new SyntaxTreeSequenceNode(position, statementNode, statementSequenceNode);
            }
            return firstStatementNode;
        }

        /*
         * statement = [ declaration | assignment | read | write ] DELIMITER_SYMBOL.
         */
        SyntaxTreeNode IsStatement()
        {
            SyntaxTreeNode statementNode = null;

            if (null != (statementNode = IsDeclaration()) ||
                null != (statementNode = IsAssignment()) ||
                null != (statementNode = IsRead()) ||
                null != (statementNode = IsWrite()))
            {
                if (this._scanner.PeekSymbol() != Symbol.DELIMITER_SYMBOL)
                {
                    return CreateErrorNode("\';\' expected");
                }
                this._scanner.NextSymbol();
                return statementNode;
            }
            else if (IsEofSymbol())
            {
                return null;
            }

            return CreateErrorNode("Statement expected");
            throw new Exception("The method or operation is not implemented.");
        }

        /*
         * declaration = typeName identifier { COMMA_SYMBOL identifier }.
         */
        SyntaxTreeNode IsDeclaration()
        {
            bool commaOccurred = false;
            SyntaxTreeNode firstDeclSequeceNode = null;
            SyntaxTreeNode declSequenceNode = null;
            SyntaxTreeNode declNode = null;
            SyntaxTreeNode typeNode;

            FilePosition position = GetInputFilePosition();
            if (null != (typeNode = IsTypeName()))
            {
                do
                {
                    Symbol currentSymbol = this._scanner.PeekSymbol();
                    if (this._scanner.SymbolTable.IsIdentifierSymbol(currentSymbol))
                    {
                        declNode = new SyntaxTreeDeclarationNode(position, currentSymbol);
                        this._scanner.NextSymbol();
                        if (this._scanner.PeekSymbol() == Symbol.EOF_SYMBOL)
                        {
                            return CreateErrorNode("Unexpected EOF", false);
                        }
                    }
                    else
                    {
                        return CreateErrorNode("Identifier expected after type name");
                    }

                    if (this._scanner.PeekSymbol() == Symbol.COMMA_SYMBOL)
                    {
                        position = GetInputFilePosition();
                        declSequenceNode = new SyntaxTreeSequenceNode(position, declNode, declSequenceNode);
                        if (!commaOccurred)
                        {
                            commaOccurred = true;
                            firstDeclSequeceNode = declSequenceNode;
                        }
                        this._scanner.NextSymbol();
                        if (this._scanner.PeekSymbol() == Symbol.EOF_SYMBOL)
                        {
                            return CreateErrorNode("Unexpected EOF", false);
                        }
                    }
                    else if (this._scanner.PeekSymbol() == Symbol.DELIMITER_SYMBOL)
                    {
                        position = GetInputFilePosition();
                        declSequenceNode = new SyntaxTreeSequenceNode(position, declNode, declSequenceNode);
                        if (!commaOccurred)
                        {
                            firstDeclSequeceNode = declSequenceNode;
                        }
                    }
                    else
                    {
                        return CreateErrorNode("\',\' or \';\' expected after an identifier in a declaration");
                    }
                } while (Symbol.DELIMITER_SYMBOL != this._scanner.PeekSymbol());
            }

            return firstDeclSequeceNode;
        }

        /*
         * assignment = identifier ASSIGN_SYMBOL intExpr.
         */
        SyntaxTreeNode IsAssignment()
        {
            SyntaxTreeNode identNode = null;
            SyntaxTreeNode intExprNode = null;

            if(null != (identNode = IsIdentifier()))
            {
                FilePosition position = GetInputFilePosition();
                if (this._scanner.PeekSymbol() == Symbol.ASSIGN_SYMBOL)
                {
                    this._scanner.NextSymbol();
                    if (null != (intExprNode = IsIntExpr()))
                    {
                        return new SyntaxTreeAssignmentNode(position, identNode, intExprNode);
                    }
                    else
                    {
                        return CreateErrorNode("IntExpr expected on right hand side of \':=\'");
                    }
                }
                else
                {
                    return CreateErrorNode("\':=\' expected");
                }
            }
            return null;
        }

        /*
         * identifier = ANY_LETTER { ANY_DIGIT | ANY_LETTER }.
         */
        SyntaxTreeNode IsIdentifier()
        {
            FilePosition position = GetInputFilePosition();
            Symbol currentSymbol = this._scanner.PeekSymbol();
            if (this._scanner.SymbolTable.IsIdentifierSymbol(currentSymbol))
            {
                this._scanner.NextSymbol();
                return new SyntaxTreeIdentNode(position, currentSymbol);
            }
            return null;
        }

        /*
         * read = READ_SYMBOL identifier.
         */
        SyntaxTreeNode IsRead()
        {
            SyntaxTreeNode identNode = null;

            FilePosition position = GetInputFilePosition();
            if (this._scanner.PeekSymbol() == Symbol.READ_SYMBOL)
            {
                this._scanner.NextSymbol();
                if (null != (identNode = IsIdentifier()))
                {
                    return new SyntaxTreeReadNode(position, identNode);
                }
                else
                {
                    return CreateErrorNode("Identifier expected after READ");
                }
            }
            return null;
        }

        /*
         * write = WRITE_SYMBOL identifier.
         */
        SyntaxTreeNode IsWrite()
        {
            SyntaxTreeNode identNode = null;
            FilePosition position = GetInputFilePosition();
            if (this._scanner.PeekSymbol() == Symbol.WRITE_SYMBOL)
            {
                this._scanner.NextSymbol();
                if (null != (identNode = IsIdentifier()))
                {
                    return new SyntaxTreeWriteNode(position, identNode);
                }
                else
                {
                    return CreateErrorNode("Identifier expected after WRITE");
                }
            }
            return null;
        }

        /*
         * intExpr = [ MINUS_SYMBOL ] term { addOp term }.
         */
        SyntaxTreeNode IsIntExpr()
        {
            bool minusFlag;
            SyntaxTreeNode leftTermNode = null;
            SyntaxTreeNode rightTreeNode = null;
            FilePosition position = GetInputFilePosition();

            Symbol currentSymbol = this._scanner.PeekSymbol();
            if (currentSymbol == Symbol.MINUS_SYMBOL)
            {
                this._scanner.NextSymbol();
                minusFlag = true;
            }
            else
            {
                minusFlag = false;
            }

            if (null == (leftTermNode = IsIntTerm()))
            {
                if (!minusFlag)
                    return null;
                else
                {
                    return CreateErrorNode("intTerm expected after \'-\'");
                }
            }

            if (minusFlag)
            {
                leftTermNode = new SyntaxTreeMonadicOpNode(position, Symbol.MINUS_SYMBOL, leftTermNode);
            }

            currentSymbol = this._scanner.PeekSymbol();
            while (currentSymbol == Symbol.MINUS_SYMBOL || currentSymbol == Symbol.PLUS_SYMBOL)
            {
                position = GetInputFilePosition();
                this._scanner.NextSymbol();
                if (null == (rightTreeNode = IsIntTerm()))
                {
                    return CreateErrorNode("intTerm expected after \'addOp\'");
                }
                else
                {
                    leftTermNode = new SyntaxTreeDyadicOpNode(position, currentSymbol, leftTermNode, rightTreeNode);
                    currentSymbol = this._scanner.PeekSymbol();
                }
            }

            return leftTermNode;
        }

        /*
         * intTerm = isIntTerm { multOp intFactor }.
         */
        SyntaxTreeNode IsIntTerm()
        {
            SyntaxTreeNode leftFactorNode = null;
            SyntaxTreeNode rightFactorNode = null;

            if (null == (leftFactorNode = IsIntFactor()))
                return null;

            Symbol currentSymbol = this._scanner.PeekSymbol();
            while (currentSymbol == Symbol.TIMES_SYMBOL ||
                currentSymbol == Symbol.DIVIDE_SYMBOL ||
                currentSymbol == Symbol.MODULO_SYMBOL)
            {
                FilePosition position = GetInputFilePosition();
                this._scanner.NextSymbol();
                if (null == (rightFactorNode = IsIntFactor()))
                {
                    return CreateErrorNode("intFactor expected after \'multOp\'");
                }
                else
                {
                    leftFactorNode = new SyntaxTreeDyadicOpNode(position, currentSymbol, leftFactorNode, rightFactorNode);
                    currentSymbol = this._scanner.PeekSymbol();
                }
            }
            return leftFactorNode;
        }

        /*
         * intFactor = integer | identifier | OPEN_PARAENTHESIS_SYMBOL intExpr CLOSE_PARENTHESIS_SYMBOL | inlineIfStatement.
         */
        SyntaxTreeNode IsIntFactor()
        {
            SyntaxTreeNode intFactorNode = null;
            SyntaxTreeNode intExprNode = null;
            FilePosition position = GetInputFilePosition();
            if (null != (intFactorNode = IsInteger()))
            {
                return intFactorNode;
            }
            else if(null != (intFactorNode = IsIdentifier()))
            {
                return intFactorNode;
            }
            else if (this._scanner.PeekSymbol() == Symbol.OPEN_PARAENTHESIS_SYMBOL)
            {
                this._scanner.NextSymbol();
                Symbol currentSymbol = this._scanner.PeekSymbol();
                if (null != (intExprNode = IsIntExpr()))
                {
                    currentSymbol = this._scanner.PeekSymbol();
                    if (currentSymbol == Symbol.CLOSE_PARENTHESIS_SYMBOL)
                    {
                        this._scanner.NextSymbol();
                        return new SyntaxTreeNodeParenthesisNode(position, intExprNode);
                    }
                    else
                    {
                        return CreateErrorNode("\')\' expected");
                    }
                }
                else
                {
                    return CreateErrorNode("intExpr expected after \'(\'");
                }
            }
            else if (null != (intFactorNode = IsInlineIf()))
            {
                return intFactorNode;
            }

            return CreateErrorNode("intFactor expected");
        }

        /*
         * integer = ANY_DIGIT { ANY_DIGIT }.
         */
        SyntaxTreeNode IsInteger()
        {
            SyntaxTreeNode constNode = null;
            FilePosition position = GetInputFilePosition();
            Symbol currentSymbol = this._scanner.PeekSymbol();
            if (this._scanner.SymbolTable.IsNumberSymbol(currentSymbol))
            {
                this._scanner.NextSymbol();
                constNode = this._scanner.SymbolTable.GetDeclarationNodeLinkToSymbol(currentSymbol);
                if (null == constNode)
                {
                    constNode = new SyntaxTreeConstNode(position, currentSymbol);
                    this._scanner.SymbolTable.SetDeclarationNodeLinkToSymbol(currentSymbol, constNode);
                }
                return constNode;
            }
            return null;
        }

        /*
         * inlineIfStatement = INLINE_IF_SYMBOL boolExpr INLINE_THEN_SYMBOL intExpr INLINE_ELSE_SYMBOL intExpr INLINE_FI_SYMBOL.
         */
        SyntaxTreeNode IsInlineIf()
        {
            SyntaxTreeNode boolExprNode = null;
            SyntaxTreeNode thenIntExprNode = null;
            SyntaxTreeNode elseIntExprNode = null;
            FilePosition position = GetInputFilePosition();

            if (this._scanner.PeekSymbol() != Symbol.INLINE_IF_SYMBOL)
            {
                return null;
            }

            this._scanner.NextSymbol();

            if (null == (boolExprNode = IsBoolExpr()))
            {
                return CreateErrorNode("boolExpr expected");
            }

            if (this._scanner.PeekSymbol() != Symbol.INLINE_THEN_SYMBOL)
            {
                return CreateErrorNode("\'?\' expected");
            }

            this._scanner.NextSymbol();

            if (null == (thenIntExprNode = IsIntExpr()))
            {
                return CreateErrorNode("intExpr expected");
            }

            if (this._scanner.PeekSymbol() != Symbol.INLINE_ELSE_SYMBOL)
            {
                return CreateErrorNode("\':\' expected");
            }

            this._scanner.NextSymbol();

            if (null == (elseIntExprNode = IsIntExpr()))
            {
                return CreateErrorNode("intExpr expected");
            }

            if (this._scanner.PeekSymbol() == Symbol.INLINE_FI_SYMBOL)
            {
                this._scanner.NextSymbol();
                return new SyntaxTreeInlineIfNode(position, boolExprNode, thenIntExprNode, elseIntExprNode);
            }

            return CreateErrorNode("\'FII\' expected");
        }

        /*
         * boolExpr = intExpr relationOp intExpr { boolOp intExpr relationOp intExpr }.
         */
        SyntaxTreeNode IsBoolExpr()
        {
            SyntaxTreeNode leftIntExprNode = null;
            SyntaxTreeNode rightIntExprNode = null;
            SyntaxTreeNode newBoolExprNode1 = null;

            if (null == (leftIntExprNode = IsIntExpr()))
            {
                return null;
            }

            Symbol currentSymbol = this._scanner.PeekSymbol();
            FilePosition position = GetInputFilePosition();
            if (!IsRelationOpSymbol(currentSymbol))
            {
                return CreateErrorNode("\'relationOp\' expected after \'intExpr\'");
            }

            this._scanner.NextSymbol();
            if (null == (rightIntExprNode = IsIntExpr()))
            {
                return CreateErrorNode("\'intExpr\' expected after \'relationOp\'");
            }

            newBoolExprNode1 = new SyntaxTreeDyadicOpNode(position, currentSymbol, leftIntExprNode, rightIntExprNode);

            Symbol currentSymbolForBoolOp = this._scanner.PeekSymbol();
            while (currentSymbolForBoolOp == Symbol.AND_SYMBOL || currentSymbolForBoolOp == Symbol.OR_SYMBOL)
            {
                position = GetInputFilePosition();
                this._scanner.NextSymbol();

                if (null == (leftIntExprNode = IsIntExpr()))
                {
                    return CreateErrorNode("\'intExpr\' expected after \'boolOp\'");
                }
                currentSymbol = this._scanner.PeekSymbol();
                if (!IsRelationOpSymbol(currentSymbol))
                {
                    return CreateErrorNode("\'relationOp\' expected after \'intExpr\'");
                }
                this._scanner.NextSymbol();
                if (null == (rightIntExprNode = IsIntExpr()))
                {
                    return CreateErrorNode("\'intExpr\' expected after \'relationOp\'");
                }
                SyntaxTreeNode newBoolExprNode2 = new SyntaxTreeDyadicOpNode(position, currentSymbol, leftIntExprNode, rightIntExprNode);
                newBoolExprNode1 = new SyntaxTreeDyadicOpNode(position, currentSymbolForBoolOp, newBoolExprNode1, newBoolExprNode2);
                currentSymbolForBoolOp = this._scanner.PeekSymbol();
            }
            return newBoolExprNode1;
        }

        /*
         * typeName = INT_TYPE_SYMBOL.
         */
        SyntaxTreeNode IsTypeName()
        {
            SyntaxTreeNode typeNode = null;
            FilePosition position = GetInputFilePosition();

            if (Symbol.INT_TYPE_SYMBOL == this._scanner.PeekSymbol())
            {
                typeNode =  new SyntaxTreeTypeNode(position, Symbol.INT_TYPE_SYMBOL);
                this._scanner.NextSymbol();
            }
            return typeNode;
        }

        bool IsEofSymbol()
        {
            return Symbol.EOF_SYMBOL == this._scanner.PeekSymbol();
        }

        bool IsRelationOpSymbol(Symbol symbol)
        {
            return
                symbol == Symbol.LT_SYMBOL ||
                symbol == Symbol.LE_SYMBOL ||
                symbol == Symbol.EQ_SYMBOL ||
                symbol == Symbol.GE_SYMBOL ||
                symbol == Symbol.GT_SYMBOL ||
                symbol == Symbol.NE_SYMBOL;
        }


        void SyntaxError(string message)
        {
            FilePosition position = GetInputFilePosition(); ;
            Logger.Error("Syntax Error near line " + position.Line + ", column " + position.Column + ": " + message);
        }

        SyntaxTreeNode CreateErrorNode(string message, bool skipToDelimiter = true)
        {
            SyntaxError(message);
            FilePosition position = GetInputFilePosition();
            if(skipToDelimiter)
                this._scanner.SkipToDelimiter();
            return new SyntaxTreeErrorNode(position);
        }
    }
}

File: NASE-Erik/src/compiler/parsers/Parser.java
package compiler.parsers;

import symboltable.SymbolTable;
import symboltable.Symbols;
import syntaxtree.nodes.AssignmentNode;
import syntaxtree.nodes.ConstantNode;
import syntaxtree.nodes.DeclarationNode;
import syntaxtree.nodes.DyadicOpNode;
import syntaxtree.nodes.ErrorNode;
import syntaxtree.nodes.IdentifierNode;
import syntaxtree.nodes.InlineIfNode;
import syntaxtree.nodes.MondanicOpNode;
import syntaxtree.nodes.Node;
import syntaxtree.nodes.ParenthesisNode;
import syntaxtree.nodes.ProgrammNode;
import syntaxtree.nodes.ReadNode;
import syntaxtree.nodes.SequenceNode;
import syntaxtree.nodes.TypeNameNode;
import syntaxtree.nodes.WriteNode;

import compiler.scanner.ScannerWrapper;

public class Parser extends ParserWrapper{
		
	public Parser(ScannerWrapper scanner) {
		this.scanner = scanner;
	}
	
	boolean isEOFSymbol(){
		return Symbols.EOF == scanner.getCurrentSymbol();
	}
	
	void reportGeneralSyntaxError(int line, int column, String error){
		listing.write(String.format("\nSYNTAX ERROR near line %d, column %d: %s", line, column, error));	
	}
	
	boolean isRelationalOp(int sym){
		return (sym == Symbols.LT_SYMBOL
				 || sym == Symbols.LE_SYMBOL
				 || sym == Symbols.EQ_SYMBOL
				 || sym == Symbols.GT_SYMBOL
				 || sym == Symbols.GE_SYMBOL
				 || sym == Symbols.NE_SYMBOL);
	}
	
	Node getErrorNode(String message){
		int column = scanner.getColumn();
		int row = scanner.getLine();
		reportGeneralSyntaxError(row, column, message);
		scanner.skipToDelimiter();
		return new ErrorNode(row, column);
	}
	
	/*
	 * typeName = INT_TYPE_SYMBOL.
	 */
	Node isTypeName(){
		Node type = NULLNODE;
		int column = scanner.getColumn();
		int row = scanner.getLine();		
		
		switch(scanner.getCurrentSymbol()){
			case Symbols.INT_TYPE_SYMBOL:
				type = new TypeNameNode(column, row, scanner.getCurrentSymbol());
				scanner.getNextSymbol();
				break;					
		}
		
		return type;
	}
	
	/*
	 * typeName = INT_TYPE_SYMBOL.
	 */
	Node isInteger(){
		int symbol = scanner.getCurrentSymbol();
		int line = scanner.getLine();
		int column = scanner.getColumn();
		
		if(SymbolTable.getInstance().isAnyNumericSymbol(symbol)){
			scanner.getNextSymbol();
			Node constantNode = SymbolTable.getInstance().getNodeLink(symbol);
			if(constantNode == null){
				constantNode = new ConstantNode(line, column, symbol);
				SymbolTable.getInstance().setNodeLink(symbol, constantNode);
			}
			return constantNode;
		}
		return NULLNODE;
	}
		
	/*
	 * declaration = typeName identifier { COMMA_SYMBOL identifier }.
	 */
	Node isDeclaration(){
		Node type;
		Node decl;
		Node declseq = NULLNODE;
		Node firstdeclseq = NULLNODE;
		
		boolean commaOccured = false;
		int column = scanner.getColumn();
		int row = scanner.getLine();		
		
		if(!NULLNODE.equals(type = isTypeName())){
			do{
				int currentSymbol = scanner.getCurrentSymbol();
				
				if(SymbolTable.getInstance().isAnyIdentifierSymbol(currentSymbol)){
					decl = new DeclarationNode(column, row, currentSymbol, type);
					if(!scanner.getNextSymbol())
						return getErrorNode("Unexpected EOF");
				} else 
					return getErrorNode("Identifier expected after type name");
				
				column = scanner.getColumn();
				row = scanner.getLine();
				
				if(scanner.getCurrentSymbol() == Symbols.COMMA_SYMBOL ){
					declseq = new SequenceNode(NULLNODE, row, column);
					declseq.addChild(decl);
					if(!commaOccured){
						commaOccured = true;
						firstdeclseq = declseq;
					}
					if(!scanner.getNextSymbol())						
						return getErrorNode("Unexpected EOF");
					
				} else if(scanner.getCurrentSymbol() == Symbols.DELIMITER_SYMBOL ){
					declseq = new SequenceNode(NULLNODE, row, column);
					declseq.addChild(decl);
					if(!commaOccured)
						firstdeclseq = declseq;
				} else 
					return getErrorNode("',' or ';' expected after an identifier declaration");	
				
			}while(scanner.getCurrentSymbol() != Symbols.DELIMITER_SYMBOL );
			return firstdeclseq;
		}
		return NULLNODE;
	}
	
	/*
	 * identifier = ANY_LETTER { ANY_DIGIT | ANY_LETTER }.
	 */
	Node isIdentifier(){
		int symbol = scanner.getCurrentSymbol();
		if(SymbolTable.getInstance().isAnyIdentifierSymbol(symbol)){
			scanner.getNextSymbol();
			return new IdentifierNode(scanner.getLine(),scanner.getColumn(), symbol);
		}
		return NULLNODE;
	}

	/*
	 * intFactor = integer | identifier | OPEN_PARAENTHESIS_SYMBOL intExpr CLOSE_PARENTHESIS_SYMBOL | inlineIfStatement.
	 */
	Node isIntFactor(){
		int symbol = scanner.getCurrentSymbol();
		int line = scanner.getLine();
		int column = scanner.getColumn();
		
		Node factor;
		Node expr;
		
		if(!NULLNODE.equals(factor = isInteger())
			 || !NULLNODE.equals(factor = isIdentifier())
			 || !NULLNODE.equals(factor = isInlineIf())){
			return factor;
		} else if(scanner.getCurrentSymbol() == Symbols.OPEN_PARAENTHESIS_SYMBOL) {
			scanner.getNextSymbol();
			symbol = scanner.getCurrentSymbol();
			if(!NULLNODE.equals(expr = isIntExpr())){
				symbol = scanner.getCurrentSymbol();
				if(symbol == Symbols.CLOSE_PARENTHESIS_SYMBOL){
					scanner.getNextSymbol();
					return new ParenthesisNode(line, column, expr);
				} else
					return getErrorNode("')' expected");
			} else
				return getErrorNode("intExpr expected after '('");
		}
		return getErrorNode("intFactor expected");
	}

	/*
	 * intTerm = isIntTerm { multOp intFactor }.
	 */
	Node isIntTerm(){
		Node leftFactor;
		Node rightFactor;
		
		if(NULLNODE.equals(leftFactor = isIntFactor()))
			return NULLNODE;
		
		int symbol = scanner.getCurrentSymbol();
	
		while(Symbols.TIMES_SYMBOL  == symbol
				|| Symbols.DIVIDE_SYMBOL  == symbol
				|| Symbols.MODULO_SYMBOL  == symbol){
			
			int column = scanner.getColumn();
			int row = scanner.getLine();			
			scanner.getNextSymbol();
			
			if(NULLNODE.equals(rightFactor = isIntFactor()))
				return getErrorNode("intFactor expected after 'multOp'");	
			else {
				Node newTerm = new DyadicOpNode(row, column, symbol, leftFactor, rightFactor);
				leftFactor = newTerm;
				symbol = scanner.getCurrentSymbol();
			}
		}
		return leftFactor;
	}
	
	/*
	 * intExpr = [ MINUS_SYMBOL ] term { addOp term }.
	 */
	Node isIntExpr(){
		boolean minusOccured = false;
		Node leftTerm;
		Node rightTerm;
		Node minusPrefix = null;
		
		int column, row;
		
		int columnMinus = scanner.getColumn();
		int rowMinus = scanner.getLine();
		if(scanner.getCurrentSymbol() == Symbols.MINUS_SYMBOL ){
			minusOccured = true;
			scanner.getNextSymbol();
		}
		
		if(NULLNODE.equals(leftTerm = isIntTerm())){
			if(!minusOccured)
				return NULLNODE;
			
			return getErrorNode("intTerm expected after '-'");	
		}
		
		if(minusOccured){
			minusPrefix = new MondanicOpNode(rowMinus, columnMinus, Symbols.MINUS_SYMBOL, leftTerm);
			leftTerm = minusPrefix;
		}
		
		while(Symbols.MINUS_SYMBOL  == scanner.getCurrentSymbol()
				|| Symbols.PLUS_SYMBOL  == scanner.getCurrentSymbol()){
			
			column = scanner.getColumn();
			row = scanner.getLine();
			
			scanner.getNextSymbol();
			if(NULLNODE.equals(rightTerm = isIntTerm()))
				return getErrorNode("intTerm expected after 'addOp'");	
			else {
				Node newTerm = new DyadicOpNode(row, column, scanner.getCurrentSymbol(), leftTerm, rightTerm);
				leftTerm = newTerm;				
			}
		}
		return leftTerm;
	}
	
	/*
	 * inlineIfStatement = INLINE_IF_SYMBOL boolExpr INLINE_THEN_SYMBOL intExpr INLINE_ELSE_SYMBOL intExpr INLINE_FI_SYMBOL.
	 */
	Node isInlineIf(){
		int line = scanner.getLine();
		int column = scanner.getColumn();
		Node boolExpr;
		Node thenExpr;
		Node elseExpr;

		if(scanner.getCurrentSymbol() == Symbols.INLINE_IF_SYMBOL){
			scanner.getNextSymbol();
			if(!NULLNODE.equals(boolExpr = isBoolExpr())){
				if(scanner.getCurrentSymbol() == Symbols.INLINE_THEN_SYMBOL){
					scanner.getNextSymbol();
					if(!NULLNODE.equals(thenExpr = isIntExpr())){
						scanner.getNextSymbol();
						if(scanner.getCurrentSymbol() == Symbols.INLINE_ELSE_SYMBOL){
							scanner.getNextSymbol();
							if(!NULLNODE.equals(elseExpr = isIntExpr())){
								scanner.getNextSymbol();
								if(scanner.getCurrentSymbol() == Symbols.INLINE_FI_SYMBOL)
									return new InlineIfNode(line, column, boolExpr, thenExpr, elseExpr);
								else
									return getErrorNode("FII expected");
							} else
								return getErrorNode("intExpr expected");
						} else
							return getErrorNode("':' expected");
					} else
						return getErrorNode("intExpr expected");
				} else 
					return getErrorNode("'?' expected");
			} else
				return getErrorNode("boolExpr expected");
		} else 
			return NULLNODE;
	}

	/*
	 * boolExpr = intExpr relationOp intExpr { boolOp intExpr relationOp intExpr }.
	 */
	Node isBoolExpr(){
		Node leftInt;
		Node rightInt;
		Node boolExpr1;
		Node boolExpr2;
		Node boolExpr3;
		
		int currentSymbol;
		int currentSymbolForBoolOp;
		int line;
		int column;
				
		if(!NULLNODE.equals(leftInt = isIntExpr())){
			currentSymbol = scanner.getCurrentSymbol();
			line = scanner.getLine();
			column = scanner.getColumn();
			
			if(isRelationalOp(currentSymbol)){
				scanner.getNextSymbol();
				if(!NULLNODE.equals(rightInt = isIntExpr()))
					boolExpr1 = new DyadicOpNode(line, column, currentSymbol, leftInt, rightInt);
				else
					return getErrorNode("'intExpr' expected after 'relationOp'");
			} else
				return getErrorNode("'relationOp' expected after 'intExpr'");
		} else
			return NULLNODE;
		
		currentSymbolForBoolOp = scanner.getCurrentSymbol();
		while(Symbols.AND_SYMBOL == currentSymbolForBoolOp
				|| Symbols.OR_SYMBOL == currentSymbolForBoolOp) {
			line = scanner.getLine();
			column = scanner.getColumn();
			scanner.getNextSymbol();
			if(!NULLNODE.equals(leftInt = isIntExpr())){
				currentSymbol = scanner.getCurrentSymbol();
				if(isRelationalOp(currentSymbol)){
					scanner.getNextSymbol();
					if(!NULLNODE.equals(rightInt = isIntExpr())){
						boolExpr2 = new DyadicOpNode(line, column, currentSymbol, leftInt, rightInt);
						boolExpr3 = new DyadicOpNode(line, column, currentSymbolForBoolOp, boolExpr1, boolExpr2);
						boolExpr1 = boolExpr3;
						currentSymbolForBoolOp = scanner.getCurrentSymbol();
					} else
						return getErrorNode("'intExpr' expected after 'relationOp'");
				} else
					return getErrorNode("'relationOp' expected after 'intExpr'");
			} else
				return getErrorNode("'intExpr' expected after 'boolOp'");
		}
		
		return boolExpr1;
	}

	/*
	 * write = WRITE_SYMBOL identifier.
	 */
	Node isWrite(){
		int line = scanner.getLine();
		int column = scanner.getColumn();
		
		Node expr;
		
		if(scanner.getCurrentSymbol() == Symbols.WRITE_SYMBOL ){
			scanner.getNextSymbol();
			
			if(!NULLNODE.equals(expr = isIntExpr())
					|| !NULLNODE.equals(expr = isBoolExpr()))
				return new WriteNode(line, column, expr);
			else
				return getErrorNode("intExpr or bool Expr expected after WRITE");
		}
		return NULLNODE;
	}
	
	/*
	 * read = READ_SYMBOL identifier.
	 */
	Node isRead(){
		int line = scanner.getLine();
		int column = scanner.getColumn();
		
		Node ident;
		
		if(scanner.getCurrentSymbol() == Symbols.READ_SYMBOL ){
			scanner.getNextSymbol();
			
			if(!NULLNODE.equals(ident = isIdentifier()))
				return new ReadNode(line, column, ident);
			else 
				return getErrorNode("Identifier expected after read");
		}
		return NULLNODE;
	}

	/*
	 * assignment = identifier ASSIGN_SYMBOL intExpr.
	 */
	Node isAssignment(){
		
		Node identNode;
		Node exprNode;
		
		if(! NULLNODE.equals(identNode = isIdentifier())){	
			int line = scanner.getLine();
			int column = scanner.getColumn();
			if(Symbols.ASSIGN_SYMBOL  == scanner.getCurrentSymbol()){
				scanner.getNextSymbol();
				if(! NULLNODE.equals(exprNode = isIntExpr()))
					return new AssignmentNode(line,column, identNode, exprNode);
				else
					return getErrorNode("IntExpr expected on right hand side of ':='");
			} else 
				return getErrorNode("':=' expected");	
		}
		return NULLNODE;
	}

	/*
	 * statement = [ declaration | assignment | read | write ] DELIMITER_SYMBOL.
	 */
	Node isStatement(){
		Node statement;
		
		if(!NULLNODE.equals(statement = isDeclaration())
				|| !NULLNODE.equals(statement = isAssignment())
				|| !NULLNODE.equals(statement = isRead())
				|| !NULLNODE.equals(statement = isWrite())){
			
			if(Symbols.DELIMITER_SYMBOL == scanner.getCurrentSymbol()){
				scanner.getNextSymbol();
				return statement;
			} else
				return getErrorNode("';' expected");
		}
		
		if(isEOFSymbol())
			return NULLNODE;
		
		return NULLNODE;
	}

	/*
	* statementSequence = statement { statement }.
	*/
	Node isStatementSequence(){
		int line = scanner.getLine();
		int column = scanner.getColumn();		
		boolean statementFound = false;
		Node statement = NULLNODE;
		Node statementSequence = NULLNODE;
		
		while(!NULLNODE.equals(statement = isStatement())){
			if(!statementFound){
				statementFound = true;
				statementSequence = new SequenceNode(null, line, column);
			}
			statementSequence.addChild(statement);
			line = scanner.getLine();
			column = scanner.getColumn();		
		}
		return statementSequence;
	}

	Node isProgram(){
		int line = scanner.getLine();
		int column = scanner.getColumn();
		scanner.getNextSymbol();
		
		Node statementSeq;
		
		if(NULLNODE.equals(statementSeq = isStatementSequence())){
			if(isEOFSymbol()){
				return new ProgrammNode(line, column, statementSeq);
			} else
				return getErrorNode("EOF symbol at program end missed");
		} else
			return NULLNODE;
	}


	public void parseProgramm() {
		listing.write(" ");
		
		rootNode = isProgram(); 
		
		listing.write("\n");
		listing.write("SYMBOLTABLE-DUMP\n---------------------\n");
		listing.write(SymbolTable.getInstance().toString());
		listing.write("SYNTAXTREE-DUMP\n---------------------\n");
		listing.write(rootNode.toString());
		listing.write("NODEHISTORY-DUMP\n---------------------\n");
		listing.write(dumpHistory());
		listing.write("SIMPLE-SYNTAXTREE-DUMP\n---------------------\n");
		listing.write("\n"+rootNode.dump(0));
		
	}
}